proposals/0002.md

```diff
@@ -0,0 +1,318 @@
+
+### The behavior of `Display`
+
+The `Display` implementation should briefly describe the error.
+It should never include information provided by the error returned from the `source`.
+Its output should be formatted as a single line.
+If it's a single sentence, it should start with a lowercase letter and end with no interpunction.
+These rules make it easy to print the whole chain of errors as described in the next section.
+
+Examples:
+
+`failed to load the file '/tmp/file'` **good**
+
+`failed to load the file '/tmp/file': file not found` **bad**:
+includes information provided by its `source`
+
+`Failed to load the file '/tmp/file'!` **bad**:
+a single sentence starting with a capital letter and ending with interpunction
+
+## Printing a chain of `Error`s
+
+When all the errors in a chain behave correctly, insightful and well formatted messages can be
+easily provided to the user.
+The error can be iterated by recursively calling `source` to print each layer and provide the full
+chain of errors that caused the failure.
+Due to strict `Display` implementation rules outlined in the previous section, there's a lot of
+creative freedom when it comes to chain formatting.
+The `source` traversal can be done manually with the `anyhow::Chain` iterator or with the
+very similar iterator `std::error::Chain` as soon as it's stabilized.
+
+An example of printing the chain manually:
+
+```rust
+println!("An error occured: {}", error);
+let mut parent = &error;
+while let Some(cause) = parent.source() {
+    println!("  caused by: {}", cause);
+    parent = cause;
+}
+```
+
+An example of printing the chain with `anyhow::Chain`, useful for logging:
+
+```rust
+println!("An error occured: {}", error);
+for cause in anyhow::Chain::new(&error) {
+    println!("  caused by: {}", cause);
+}
+```
+
+Both examples have the same output:
+
+```
+An error occured: failed to load the configuration file
+  caused by: failed to load the file '~/.app_config'
+  caused by: file not found
+```
+
+An example of printing the chain in a single line with `anyhow::Chain`:
+
+```rust
+print!("An error occured: {}", error);
+anyhow::Chain::new(&error).for_each(|e| print!(" ⮩ {}", e);
+println!();
+```
+
+An example output:
+
+```
+An error occured: failed to load the configuration file ⮩ failed to load the file '~/.app_config' ⮩ file not found
+```
+
+# Recommendation
+
+These examples of building errors serve as a practical guide.
+There are some common patterns of error structures and all of them can be implemented conveniently
+and correctly using `thiserror` and `anyhow`.
+
+All the examples use `enum`s, but it's fine to use `struct`s instead when there's only one variant.
+Usually it's not very useful, because it reduces flexibility, but doesn't simplify code much.
+
+## Root cause errors
+
+These are errors that don't wrap other errors inside.
+They may contain multiple structures and a lot of details, but their `source` always returns `None`.
```

Comment:
```suggestion
They may contain multiple structures and a lot of details, but they have no `source`.
```

Reply:
This would obscure their observed behavior, which is returning a specific value.